source/fundamentals/gridfs.txt

```diff
@@ -0,0 +1,387 @@
+======
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+This guide shows you how to store and retrieve large files in MongoDB using
+**GridFS**. GridFS is a specification implemented by the driver that describes
+how to split files into chunks when storing them and reassemble them when
+retrieving them. The driver implementation of GridFS is an abstraction that
+manages the operations and organization of the file storage.
+
+You should use GridFS if the size of your files exceed the BSON document
+size limit of 16MB. For more detailed information on whether GridFS is
+suitable for your use case, see the :manual:`GridFS server manual page </core/gridfs>`.
+
+See the following sections that describe GridFS operations and how to
+perform them:
+
+- :ref:`Create a GridFS bucket <gridfs-create-bucket>`
+- :ref:`Store Files <gridfs-store-files>`
+- :ref:`Retrieve File Information <gridfs-retrieve-file-info>`
+- :ref:`Download Files <gridfs-download-files>`
+- :ref:`Rename Files <gridfs-rename-files>`
+- :ref:`Delete Files <gridfs-delete-files>`
+- :ref:`Delete a GridFS bucket <gridfs-delete-bucket>`
+
+How GridFS Works
+----------------
+
+GridFS organizes files in a **bucket**, a group of MongoDB collections
+that contain the chunks of files and information describing them. The
+bucket contains the following collections, named using the convention
+defined in the GridFS specification:
+
+- The ``chunks`` collection stores the binary file chunks.
+- The ``files`` collection stores the file metadata.
+
+When you create a new GridFS bucket, the driver creates the collections
+referenced above, prefixed with the default bucket name ``fs``, unless
+you specify a different name. The driver also creates an index on each
+collection to ensure efficient retrieval of the files and related
+metadata. For more information on GridFS indexes, see the server manual
+page on :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`.
+
+When storing files with GridFS, the driver splits the files into smaller
+chunks, each represented by a separate document in the ``chunks`` collection.
+It also creates a document in the ``files`` collection that contains
+a file id as an ``ObjectId``, file name, and other file metadata. You can
+upload the file from memory or from a stream. See the following diagram to
+see how GridFS splits the files when uploaded to a bucket.
+
+.. figure:: /includes/figures/GridFS-upload.png
+   :alt: A diagram that shows how GridFS uploads a file to a bucket
+
+When retrieving files, GridFS fetches the metadata from the ``files``
+collection in the specified bucket and uses the information to reconstruct
+the file from documents in the ``chunks`` collection. You can read the file
+into memory or output it to a stream.
+
+.. _gridfs-create-bucket:
+
+Create a GridFS Bucket
+----------------------
+
+To use GridFS, first create a bucket or reference an existing one on
```

Comment:
```suggestion
To use GridFS, create a bucket or reference an existing one on
```
I'm not sure if we need "first" here since this statement basically says "to use GridFS, you need a GridFS bucket". I guess the "first" is supposed to modify both "create a bucket" and "reference an existing one", but to me it reads as if it's only modifying "create a bucket" and I was tripped up by the fact that there wasn't really another step there. Here's an idea for how you can break this up to avoid that confusion:

```
To use GridFS, you'll need a GridFS bucket. You can either create a bucket or
reference an existing one in a MongoDB database. Call the...
```